borsh/src/schema.rs

```diff
@@ -333,7 +333,7 @@ where
 #[cfg(hash_collections)]
 pub mod hashes {
     //!
-    //! Module defines [BorshSchema](crate::schema::BorshSchema) implementation for  
+    //! Module defines [BorshSchema](crate::schema::BorshSchema) implementation for
     //! [HashMap](std::collections::HashMap)/[HashSet](std::collections::HashSet).
     use crate::BorshSchema;
 
```

Comment:
replace `Definition` in this file with the one proposed.

```rust
pub enum Definition {
    /// A sequence of elements of the same type.
    Sequence {
        /// How many bytes does the length prefix occupy.
        ///
        /// Zero means this is a fixed-length array or a custom encoding schema
        /// which doesn’t follow typical borsh conventions.  In the former case,
        /// `min_length` and `max_length` will be set to the same value.  In the
        /// latter case, it may be impossible to deserialise the data without
        /// additional knowledge of the declared type.
        length_width: u8,

        /// Minimum length of the sequence.  Usually `0` for dynamically-sized
        /// sequences.
        min_length: u64,

        /// Maximum length of the sequence.  Usually `2**length_width - 1`.
        max_length: u64,

        /// Type of the elements of the sequence.
        elements: Declaration,
    },

    /// A fixed-size tuple with the length known at the compile time and the
    /// elements of different types.
    Tuple {
        /// Types of elements of the tuple.
        elements: Vec<Declaration>,
    },

    /// A tagged union, a.k.a enum. Tagged-unions have variants with associated
    /// structures.
    Enum {
        /// Width of the discriminant tag.
        tag_width: u8,

        /// Possible variants of the enumeration.
        variants: Vec<(VariantName, Declaration)>,
    },

    /// A structure, structurally similar to a tuple.
    Struct {
        fields: Fields,
    },
}
```

adjust `max_serialized_size` implementation accordingly.

Reply:
A couple of questions about this:

First of all, what’s even the point of Fields?  Specifically, why do
we bother encoding unnamed struct as a struct rather than as a tuple.
This only complicates clients of the code since they have two
identical cases to consider separately.

And secondly, (while I would even go as far as get rid of field names,
but baring that) why not get rid of Struct and instead use:

```rust
    Tuple {
        /// Types of elements of the tuple.
        elements: Vec<(Option<FieldName>, Declaration)>,
    },
```

Reply:
@mina86 let's address only `Array`, `Enum` and `Sequence` in this pr, as these are where __new__ details are being added, the proposed changes to `Struct` aren't adding new information, they are just rearranging existing representation in a slightly different manner, albeit (arguably) in a more correct one.
We can try to include another pr with `Definition::Struct/Tuple` changes into 1.0.0 too.

Reply:
@mina86 on second thought, it might be better to accept this pr as is.
And then issue 3 separate prs:
1. remove `Definition::Array` , replace it with new extended `Definition::Sequence` + a test for a local `BoundedVec<T, L, U>` `max_serialized_size`
2. add `tag_width` to `Definition::Enum`
3. make the change so that tuple structs become modelled as a `Tuple`
```rust

    /// A fixed-size tuple with the length known at the compile time and the
    /// elements of different types.
    Tuple {
        /// Types of elements of the tuple.
        elements: Vec<Declaration>,
    },
    /// A structure, structurally similar to a tuple.
    Struct {
        /// Named fields of the structure with their types.
        ///
        /// Tuple-like structures are modelled as a `Tuple`.
        fields: Vec<(FieldName, Declaration)>,
    },
```

Thus it will be easier for everyone to digest it in 4 separate chunks and do additional clarification per each point if needed. I completely skipped the part with point 3. on first pass.
What do you think?

Reply:
I’m already working on the other things as separate PRs.  Whether this PR should be accepted now or later I have no opinion.